src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -984,6 +985,10 @@ trait Contexts { self: Analyzer =>
 
     def isNameInScope(name: Name) = lookupSymbol(name, _ => true).isSuccess
 
+    def isPlainClassInJavaUnit(sym: Symbol): Boolean =
```

Comment:
To make it a little trickier:

```
// Java
public class A {
  public static class I { }
}

// Scala
class B {
  // class I  // (1)
}
class C extends A
object C extends B {
  // class I  // (2)
}
```

What is `C.I` in a Java source? In separate compilation (mixed compilation should behave the same):
  - With 1 and 2 commented, `C.I` (in a Java source) resolves to `A$I`
  - Un-commenting 1 doesn't change anything: class `B` is not a parent of the class `C` - it's a parent of the module class `C$`. If a Java source references `C.I`, the Java compiler doesn't look in `C$`.
  - Un-commenting 2 triggers some compat-mode in the `InnerClass` attribute and marks `I` as being a static member class of `C` (not of `C$`), so a `C.I` reference in Java resolves to `C$I`.

So when we are in a Java compilation unit and look for a member of a module class `C` we should
  - first check the `C` module class (but not any of its parents)
  - then the `C` class and its parents, and also all parents module classes

If we also start type checking Java terms (in the future) things would get more complicated. Right now, in `findMember`, we don't know whether we're looking for a static or instance member. A type selection `Outer.Inner`, it can resolve to a static or non-static inner class, the same syntax is used for both. For terms, `A.foo` should not return an instance method.

Related older fix I wasn't aware of: https://issues.scala-lang.org/browse/SI-3120. When type-checking a selection `A.foo` fails in a Java compilation unit, a synthetic `SelectFromTypeTree` tree will be type-checked as second try.